pkg/kubelet/cm/dra/plugin/noderesources.go

```diff
@@ -1,520 +0,0 @@
-	// where we publish it.
-	//
-	// The long-term goal is to move the handling of
-	// ResourceSlice objects into the driver, with kubelet
-	// just acting as a REST proxy. The advantage of that will
-	// be that kubelet won't need to support the same
-	// resource API version as the driver and the control plane.
-	// With that approach, the driver will be able to match
-	// up objects more intelligently.
-	numObsoleteSlices := len(obsoleteSlices)
-	for index, resource := range driverResources {
-		if storedResourceIndices.Has(index) {
-			// No need to do anything, it is already stored exactly
-			// like this in an existing slice.
-			continue
-		}
-
-		if numObsoleteSlices > 0 {
-			// Update one existing slice.
-			slice := obsoleteSlices[numObsoleteSlices-1]
-			numObsoleteSlices--
-			slice = slice.DeepCopy()
-			slice.ResourceModel = *resource
-			logger.V(5).Info("Reusing existing node resource slice", "slice", klog.KObj(slice))
-			if _, err := c.kubeClient.ResourceV1alpha2().ResourceSlices().Update(ctx, slice, metav1.UpdateOptions{}); err != nil {
-				return fmt.Errorf("update node resource slice: %w", err)
-			}
-			continue
-		}
-
-		// Although node name and UID are unlikely to change
-		// we're getting updated node object just to be on the safe side.
-		// It's a cheap operation as it gets an object from the node informer cache.
-		node, err := c.getNode()
-		if err != nil {
-			return fmt.Errorf("retrieve node object: %w", err)
-		}
-
-		// Create a new slice.
-		slice := &resourceapi.ResourceSlice{
-			ObjectMeta: metav1.ObjectMeta{
-				GenerateName: node.Name + "-" + driverName + "-",
-				OwnerReferences: []metav1.OwnerReference{
-					{
-						APIVersion: v1.SchemeGroupVersion.WithKind("Node").Version,
-						Kind:       v1.SchemeGroupVersion.WithKind("Node").Kind,
-						Name:       node.Name,
-						UID:        node.UID,
-						Controller: ptr.To(true),
-					},
-				},
-			},
-			NodeName:      node.Name,
-			DriverName:    driverName,
-			ResourceModel: *resource,
-		}
-		logger.V(5).Info("Creating new node resource slice", "slice", klog.KObj(slice))
-		if _, err := c.kubeClient.ResourceV1alpha2().ResourceSlices().Create(ctx, slice, metav1.CreateOptions{}); err != nil {
-			return fmt.Errorf("create node resource slice: %w", err)
-		}
-	}
-
-	// All remaining slices are truly orphaned.
-	for i := 0; i < numObsoleteSlices; i++ {
-		slice := obsoleteSlices[i]
-		logger.V(5).Info("Deleting obsolete node resource slice", "slice", klog.KObj(slice))
-		if err := c.kubeClient.ResourceV1alpha2().ResourceSlices().Delete(ctx, slice.Name, metav1.DeleteOptions{}); err != nil {
-			return fmt.Errorf("delete node resource slice: %w", err)
-		}
-	}
-
-	return nil
-}
-
-func indexOfModel(models []*resourceapi.ResourceModel, model *resourceapi.ResourceModel) int {
-	for index, m := range models {
-		if apiequality.Semantic.DeepEqual(m, model) {
-			return index
-		}
-	}
-	return -1
-}
```

Comment:
So, the general idea is to move all of this logic to **drivers** instead of doing it once in Kubelet?

Reply:
Yes.

> Another point is that the current approach used in the CSI and Device plugins and worked for years. Why should we do it differently with DRA?

Because it does not work *well*. With DRA it's particularly bad because the data provided by drivers is much more complex and likely to change, which leads to the [version skew problem](https://github.com/kubernetes/kubernetes/issues/123699).

Reply:
For the full discussion, see the [2024-05-22 SIG Auth meeting recording](https://youtu.be/dFzEUjxs6Z0?si=xF2ba4TvX7B5uKFZ).